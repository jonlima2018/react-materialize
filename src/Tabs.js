import React, { useEffect, useRef, cloneElement } from 'react';
import PropTypes from 'prop-types';
import idgen from './idgen';
import cx from 'classnames';

import Row from './Row';
import Tab from './Tab';

const scope = `tabs-${idgen()}`;

const Tabs = ({ children, className, defaultValue, options, onChange }) => {
  const _tabsRef = useRef(null);

  useEffect(() => {
    const instance = M.Tabs.init(_tabsRef.current, options);
    return () => instance.destroy();
  }, [options, children]);

  return (
    <React.Fragment>
      <ul className={cx('tabs', className)} ref={_tabsRef}>
        {React.Children.map(children, (child, id) => {
          const idx = child.props.idx ? child.props.idx : `${scope}${id}`;
          const { active, disabled, tabWidth, title } = child.props;

          const classes = {
            [`s${tabWidth}`]: tabWidth,
            tab: true,
            disabled,
            col: true
          };

          return (
            <li className={cx(classes)} key={idx}>
              <a
                href={`#tab_${idx}`}
                className={active || defaultValue === idx ? 'active' : ''}
                {...(disabled ? {} : { onClick: onChange })}
              >
                {title}
              </a>
            </li>
          );
        })}
      </ul>
      <Row>
        {React.Children.map(children, (child, id) => {
          const idx = child.props.idx ? child.props.idx : `${scope}${id}`;
          return cloneElement(child, { idx });
        })}
      </Row>
    </React.Fragment>
  );
};

Tabs.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  defaultValue: PropTypes.string,
  onChange: PropTypes.func,
  /**
   * More info
   * <a href='http://materializecss.com/tabs.html'>http://materializecss.com/tabs.html</a>
   */
  options: PropTypes.shape({
    /**
     * Transition duration in milliseconds.
     * @default 300
     */
    duration: PropTypes.number,
    /**
     * Callback for when a new tab content is showns.
     * @default null
     */
    onShow: PropTypes.func,
    /**
     * Set to true to enable swipeable tabs. This also uses the responsiveThreshold option.
     * @default false
     */
    swipeable: PropTypes.bool,
    /**
     * The maximum width of the screen, in pixels, where the swipeable functionality initializes.
     * @default Infinity
     */
    responsiveThreshold: PropTypes.number
  })
};

Tab.defaultProps = {
  options: {
    duration: 300,
    onShow: null,
    swipeable: false,
    responsiveThreshold: Infinity
  }
};

export default Tabs;
