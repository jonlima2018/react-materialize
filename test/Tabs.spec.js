import React from 'react';
import { shallow, mount } from 'enzyme';
import Tabs from '../src/Tabs';
import Tab from '../src/Tab';
import mocker from './helper/new-mocker';

describe('Tabs', () => {
  let wrapper;
  let wrapperWithIdx;
  const tabsInitMock = jest.fn();
  const tabInstanceDestroyMock = jest.fn();
  const tabsMock = {
    init: (el, options) => {
      tabsInitMock(options);
      return {
        destroy: tabInstanceDestroyMock
      };
    }
  };
  const restore = mocker('Tabs', tabsMock);

  const options = {
    duration: 400,
    onShow: jest.fn(),
    swipeable: true,
    responsiveThreshold: 2
  };

  wrapper = shallow(
    <Tabs>
      <Tab title="one">One</Tab>
      <Tab title="Two">Two</Tab>
    </Tabs>
  );

  wrapperWithIdx = shallow(
    <Tabs>
      <Tab idx={'tab-one'} title="one">
        One
      </Tab>
      <Tab idx={'tab-two'} title="Two">
        Two
      </Tab>
    </Tabs>
  );

  afterAll(() => {
    restore();
  });

  beforeEach(() => {
    tabsInitMock.mockClear();
  });

  test('should create list of Tab items', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should create list of Tab items with Idx', () => {
    expect(wrapperWithIdx).toMatchSnapshot();
  });

  describe('with options', () => {
    beforeEach(() => {
      wrapper = mount(
        <Tabs options={options}>
          <Tab title="one">One</Tab>
          <Tab title="Two">Two</Tab>
        </Tabs>
      );
      wrapperWithIdx = mount(
        <Tabs options={options}>
          <Tab idx={'tab-one'} title="one">
            One
          </Tab>
          <Tab idx={'tab-two'} title="Two">
            Two
          </Tab>
        </Tabs>
      );
    });

    test('initializes Tabs with options', () => {
      expect(tabsInitMock).toHaveBeenCalledWith(options);
    });
  });

  describe('when unmounted', () => {
    beforeEach(() => {
      tabInstanceDestroyMock.mockClear();
      wrapper.unmount();
      wrapperWithIdx.unmount();
    });

    test('should be destroyed', () => {
      expect(tabInstanceDestroyMock).toHaveBeenCalled();
    });
  });
});
