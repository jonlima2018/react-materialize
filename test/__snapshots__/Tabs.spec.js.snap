// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should create list of Tab items 1`] = `
<Fragment>
  <ul
    className="tabs"
  >
    <Component
      tabs={true}
    />
  </ul>
  <Row>
    <Component
      content={true}
    />
  </Row>
</Fragment>
`;

exports[`Tabs should create list of Tab items with Idx 1`] = `
<Fragment>
  <ul
    className="tabs"
  >
    <Component
      tabs={true}
    />
  </ul>
  <Row>
    <Component
      content={true}
    />
  </Row>
</Fragment>
`;
